swarm/shed/rushed/subscribe.go

```diff
@@ -0,0 +1,115 @@
+package rushed
+
+import (
+	"github.com/ethereum/go-ethereum/swarm/shed"
+)
+
+const (
+	iterBatchSize = 128
+)
+
+type Subscription struct {
+	cancel chan struct{} // cancel the subscription
+	err    error
+}
+
+func Subscribe(index *shed.Index, buffer chan *shed.IndexItem, from *shed.IndexItem, trigger chan struct{}) *Subscription {
+	cancel := make(chan struct{})
+	f := func(item *shed.IndexItem) (bool, error) {
+		select {
+		case buffer <- item:
+			return false, nil
+		case <-cancel:
+			return false, errCancelled
+		}
+	}
+	s := &Subscription{
+		cancel: cancel,
+	}
+	wait := func() (bool, error) {
+		select {
+		case <-trigger:
+			return false, nil
+		case <-cancel:
+			return false, errCancelled
+		}
+	}
+	go func() {
+		defer close(buffer)
+		s.err = Iterate(index, from, f, wait)
+	}()
+	return s
+}
+
+// iterate is a wrapper to shed.IterateFrom that periodically iterates starting from 'from'
+// and remembering the last item on each round and continue the iteration from this on the
+// following round
+// once the items are retrieved in a fixed slice of max iterBatchSize elements
+// it iterates over this slice and applies f to each element
+// f returns a bool which when true terminates the iteration
+// error returned from f result in terminating the iteration and returning the error
+// if the iterator reached the last item in the index it calls the wait function
+func Iterate(index *shed.Index, from *shed.IndexItem, f func(*shed.IndexItem) (bool, error), wait func() (bool, error)) error {
```

Comment:
Please, try to make this function easier to consume by increasing the abstraction level and making the function shorter. => Extract methods, name them well.

Note: I worked a little bit with Ruby. There is a tool there Rubocop, which is kind of a really strict, configurable linter. There the default max method length is [10 lines](https://github.com/rubocop-hq/rubocop/blob/master/config/default.yml#L1696).

Reply:
> There the default max method length is 10 lines.

That is .... crazy . Anyhow @janos is working on this part

Reply:
Exactly. Crazy. I meant that as the other side of the scale. :)
But for top level functions, 10 high level lines can mean a lot of work.